File: Cocktail Ingredient PCA.R
library(tidyverse)
library(janitor)
library(tidymodels)
library(tidytext)

df <- read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-26/boston_cocktails.csv")

df |>
  count(ingredient, sort = TRUE)

# Notice that the data is very messy

# Data cleaning ----
df <- df |>
  mutate(ingredient = str_to_lower(ingredient),
         ingredient = str_replace_all(ingredient, "-", " "),
         ingredient = str_remove(ingredient, " liqueur"),
         ingredient = str_remove(ingredient, " (if desired)"),
         ingredient = case_when(str_detect(ingredient, "bitters") ~ "bitters",
                                str_detect(ingredient, "orange") ~ "orange juice",
                                str_detect(ingredient, "lemon") ~ "lemon juice",
                                str_detect(ingredient, "lime") ~ "lime juice",
                                str_detect(ingredient, "grapefruit") ~ "grapefruit juice",
                                TRUE ~ ingredient),
         # For bitters, the unit is probably a dash and not an ounce
         measure = case_when(str_detect(ingredient, "bitters") ~ str_replace(measure, "oz$", "dash"),
                             TRUE ~ measure),
         # Convert factions to decimals
         measure = str_replace(measure, " ?1/2", ".5"),
         measure = str_replace(measure, " ?1/4", ".25"),
         measure = str_replace(measure, " ?3/4", ".75"),
         measure_number = parse_number(measure),
         # If the measure is a dash, call it 1/50 of an ounce
         measure_number = if_else(str_detect(measure, "dash"), measure_number / 50, measure_number)) |>
  add_count(ingredient) |>
  filter(n >= 15) |>
  select(-n) |>
  distinct(row_id, ingredient, .keep_all = TRUE) |>
  na.omit()

# Pivot measure numbers based on ingredient and clean names
df <- df |>
  select(-ingredient_number, -row_id, -measure) |>
  pivot_wider(names_from = ingredient, values_from = measure_number, values_fill = 0) |>
  clean_names()

# The data is now much cleaner, albeit not perfect

# PCA ----
pca_recipe <- recipe(~ ., data = df) |>
  # name + category = id
  update_role(name, category, new_role = "id") |>
  step_normalize(all_predictors()) |>
  step_pca(all_predictors())

# Perform the PCA
pca_prep <- prep(pca_recipe)

# Tidy the results
pca_tidied <- tidy(pca_prep, 2)

# Look at the contributions of each ingredient to the first 5 components
pca_tidied |>
  filter(component %in% paste0("PC", 1:5)) |>
  mutate(component = fct_inorder(component)) |>
  ggplot(mapping = aes(x = value, y = terms, fill = terms)) +
  geom_col(show.legend = FALSE) +
  facet_wrap(~component, nrow = 1) +
  labs(y = NULL)

# Look at the top 8 contributors to each of the first 4 components
pca_tidied |>
  filter(component %in% paste0("PC", 1:4)) |>
  group_by(component) |>
  top_n(8, abs(value)) |>
  ungroup() |>
  mutate(terms = reorder_within(terms, abs(value), component)) |>
  ggplot(mapping = aes(x = abs(value), y = terms, fill = value > 0)) +
  geom_col() +
  scale_y_reordered() +
  facet_wrap(~component, scales = "free_y") +
  labs(y = NULL, fill = "Positive?")

# Where are the name and categories are in the 5-D space?
juice(pca_prep)

juice(pca_prep) |>
  ggplot(mapping = aes(PC1, PC2, label = name)) +
  geom_point(mapping = aes(color = category)) +
  geom_text(check_overlap = TRUE, hjust = "inward") +
  labs(color = NULL)

File: Cocktail Ingredient PCA.py
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA

def read_boston_cocktails_data():
    df = pd.read_csv("https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-26/boston_cocktails.csv")
    return df

def clean_data(df):
    # Take a look at the data to see what needs to be cleaned
    # print(df)

    # Convert ingredient to lowercase and replace hyphens with spaces
    df["ingredient"] = df["ingredient"].str.lower().str.replace("-", " ")

    # Replace " liqueur" and " (if desired)" with
    df["ingredient"] = df["ingredient"].str.replace(" liqueur", "").str.replace(" \\(if desired\\)", "")

    # Simplify the categories (for example, anything containing "bitters" just becomes "bitters")
    df["ingredient"] = np.select(
        [
            df["ingredient"].str.contains("bitters"),
            df["ingredient"].str.contains("orange"),
            df["ingredient"].str.contains("lemon"),
            df["ingredient"].str.contains("lime"),
            df["ingredient"].str.contains("grapefruit")
        ],
        [
            "bitters",
            "orange juice",
            "lemon juice",
            "lime juice",
            "grapefruit juice"
        ],
        default=df["ingredient"]
    )

    return df

def parse_measure_numbers(df):
    # For bitters, replace "oz" with "dash" in measure
    df.loc[df["ingredient"] == "bitters", "measure"] = df["measure"].str.replace("oz$", "dash")

    # Convert fractions to decimals in measure
    df["measure"] = df["measure"].str.replace(" ?1/2", ".5").str.replace(" ?1/4", ".25").str.replace(" ?3/4", ".75")

    # Parse measure_number from measure
    df["measure_number"] = df["measure"].str.extract("(\\d*\\.?\\d+)", expand=False).astype(float)

    # If the measure is a dash, consider it as 1/50 of an ounce
    df.loc[df["measure"].str.contains("dash"), "measure_number"] /= 50

    return df

def shape_top_ingredients(df):
    # Add count column for each ingredient
    df["n"] = df.groupby("ingredient")["ingredient"].transform("count")

    # Filter rows with count greater than or equal to 15
    df = df[df["n"] >= 15]

    # Select relevant columns and keep distinct rows based on row_id and ingredient
    df = (df[["name", "category", "row_id", "ingredient_number", "ingredient", "measure", "measure_number"]].
          drop_duplicates(subset=["row_id", "ingredient"], keep="first"))

    # Drop rows with missing values
    df = df.dropna()

    # Select relevant columns
    df = df.drop(columns=['ingredient_number', 'row_id', 'measure'])

    # Pivot the DataFrame wider
    df = df.pivot_table(index=df.name, columns='ingredient', values='measure_number', fill_value=0)

    # Clean column names
    df.columns = df.columns.str.lower().str.replace(' ', '_')

    return df

def perform_pca(df):
    pca = PCA(n_components=0.9).fit(df)
    print("95% of the variance:", pca.explained_variance_ratio_)

    df = pd.DataFrame(pca.components_, columns=df.columns).T

    return df

def main():
    df = read_boston_cocktails_data()

    df = clean_data(df)
    df = parse_measure_numbers(df)
    df = shape_top_ingredients(df)

    df = perform_pca(df)

if __name__ == "__main__":
    main()
